### DISCUSSION TITLE
Specifying explicit type on array column in Postgres results in 'malformed array literal' error

### DISCUSSION THREAD

**Discussion by evenfrost** [mikro-orm/mikro-orm#2041] [Q&A] [answered]
**Describe the bug**
In PostgreSQL, when you define columns with explicit `columnType`, insert operations generate invalid syntax (`["value"]` instead of `{value}`), which results in `malformed array literal: "["value"]"` error. If column type is not specified and the type inherited from TS, insert queries are generated properly.

**Stack trace**
```
(node:479180) UnhandledPromiseRejectionWarning: DriverException: insert into "dashboardHealthCheckItems" ("category", "createdAt", "details", "id", "impact", "label", "prerequisites", "product", "solution", "title", "type", "updatedAt", "valueDescription", "valueLabels") values ('branding', '2021-07-15T08:47:36.511Z', '', 'd86b569f-0c28-4baa-b27a-2c91e8a165fe', 'low', 'missing-avatars', '["foo"]', 'zendesk-support', '', '', 'suggestion', '2021-07-15T08:47:36.512Z', '', '["customer-friendliness","account-branding"]') returning "id" - malformed array literal: "["foo"]"
at PostgreSqlExceptionConverter.convertException (/home/evenfrost/work/projects/node_modules/@mikro-orm/core/platforms/ExceptionConverter.js:8:16)
at PostgreSqlExceptionConverter.convertException (/home/evenfrost/work/projects/node_modules/@mikro-orm/postgresql/PostgreSqlExceptionConverter.js:40:22)
at PostgreSqlDriver.convertException (/home/evenfrost/work/projects/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:194:54)
at /home/evenfrost/work/projects/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:198:24
at processTicksAndRejections (internal/process/task_queues.js:95:5)
at PostgreSqlDriver.nativeInsert (/home/evenfrost/work/projects/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:154:21)
at ChangeSetPersister.persistNewEntity (/home/evenfrost/work/projects/node_modules/@mikro-orm/core/unit-of-work/ChangeSetPersister.js:55:21)
at ChangeSetPersister.executeInserts (/home/evenfrost/work/projects/node_modules/@mikro-orm/core/unit-of-work/ChangeSetPersister.js:24:13)
at UnitOfWork.commitCreateChangeSets (/home/evenfrost/work/projects/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:501:9)
at UnitOfWork.persistToDatabase (/home/evenfrost/work/projects/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:458:13)


...
```

**To Reproduce**
1. Define your entity with array column and explicit `columnType` (`type: ArrayType` causes the error too).
```ts
  @Property({ columnType: 'text[]' })
  tags: string[];
```
2. Run a simple insert query, e.g.
```ts
const testRepository = em.getRepository(TestEntity);
testRepository.create({ tags: ['foo'] });
await testRepository.persistAndFlush();
```

**Expected behavior**
Query should complete successfully. Instead, `malformed array literal: "["foo"]"` error will be thrown.

**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | 14.17.0 |
| typescript | 4.3.5 |
| mikro-orm | 4.5.7 |
| your-driver | PostgreSQL |

**Comment by B4nan** [comment, marked as answer]
You need to mark the column as array, specifying `columnType` is not enough, it is actually harmful here as the discovery will skip automatic detection of the type.

```ts
@Property({ columnType: 'text[]', type: ArrayType })
tags: string[];
```

But you don't really need this, it is equivalent to:

```ts
@Property()
tags: string[];
```

As with reflect metadata we will see there is an array and by default the array is an array of strings, mapped to `text[]` column in postgres.

You can also specify the runtime type, that way the ORM will also distinguish numeric arrays:

```ts
@Property({ type: 'string[]' }) // or type: 'number[]' for numeric arrays
tags: string[];
```

> (type: ArrayType causes the error too).

Nope, it works just fine with the `ArrayType` - verified in the tests just now. That is the essence of this issue - when you specify the `columnType`, discovery won't automatically use the `ArrayType` for such column. And `ArrayType` is where the conversion to pg arrays (`{...}`) happens.

**Reply by evenfrost** [reply]
Seems I checked `type: ArrayType` without adding `columnType: 'text[]'`, and it's what caused the issue for `ArrayType`. Thanks for the help, I'm defining the array columns with the simple `@Property()` format at the moment.

**Reply by B4nan** [reply]
`@Property({ type: ArrayType })` works fine for me too, the default column type is then used from the platform, which again for pg is `text[]` (see https://github.com/mikro-orm/mikro-orm/blob/master/packages/postgresql/src/PostgreSqlPlatform.ts#L74)

It should be probably safe to also take the provided `columnType` into account and check if it ends with `[]`, will try to make it work like that in future versions.
